text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,266 @@
+
+### SSR placeholders
+
+Implementation of placeholder components, as described in [Exception A to the happy path](#the-happy-path) should require no extra effort, so long as unexpected subtree mismatches (Exception C) are dealt with properly.
+
+Implementation in user-land is possible with simple runtime environment detection:
+
+<details open>
+  <summary>Example:</summary>
+
+```javascript
+import { LightningElement } from 'lwc';
+import { isNodeEnv } from './my-utils';
+import tmplServerPlaceholder from './templateServer.html';
+import tmplClient from './templateClient.html';
+
+export default class ElementWithSSRPlaceholder extends LightningElement {
+    render() {
+        return this.isNodeEnv ? tmplServerPlaceholder : tmplClient;
+    }
+}
+```
+</details><br />
+
+Alternately, if it is desirable _not_ to ship the placeholder in the client-side bundle, one could utilize a Babel transform or `@rollup/plugin-replace` to make the necessary changes at build time.
+
+<details open>
+  <summary>Example:</summary>
+
+```javascript
+import { LightningElement } from 'lwc';
+import { isNodeEnv } from './my-utils';
+// ↶ removed from client builds during minification
+import tmplServerPlaceholder from './templateServer.html';
+import tmplClient from './templateClient.html';
+
+export default class ElementWithSSRPlaceholder extends LightningElement {
+    render() {
+        return process.browser
+            ? tmplServerPlaceholder
+            : tmplClient;
+        /*
+          For client builds, further minification iterations will result in:
+            1. return false ? tmplServerPlaceholder : tmplClient
+            2. return tmplClient
+        */
+    }
+}
+```
+
+</details>
+
+## Alternatives
+
+As detailed below in the [Survey of Prior Art](#survey-of-prior-art), some web frameworks choose _not_ to be resilient to unexpected changes in state and the resulting HTML/VDOM. For example, React > 16 will no longer protect against unexpected mismatched in production mode. However, React _also_ considers complex state management out of scope for the library.
+
+For performance reasons, Vue will also not repair mismatching HTML during rehydration. However, this is less of an issue in practice, because Vue will take a snapshot of state and ship it alongside the rendered application as `window.__INITIAL_STATE__`.
+
+Most framework authors have indicated that non-support for rehydration resilience is due to performance reasons. We should measure the performance impact of SSR rehydration in a large LWC app to either 1) validate our design decision, or 2) point us in a new direction.
+
+Additionally, if we determine that the performance impact is too great, a solution will need to be devised to address [Exception A to the happy path](#the-happy-path).
+
+# Unresolved questions
+
+- **How do we handle invocations of `customElements.define` with SSR-serialized custom elements already in the DOM?**

```

Comment:
The whole invariant around what happens with dynamic or lazy custom elements should be resolved. At least define an ambiguous behaviour